### DISCUSSION TITLE
What kind of utils functions can be in the code of airflow (providers/core)?

### DISCUSSION THREAD

**Discussion by JavierLopezT** [apache/airflow#16652] [General] [answered]
Hello. I have some utils function that I use and I was wondering whether or not they fit for being in the code. I mean functions like:

copy_firehose_files_to_airflow_bucket (aws.utils.s3?)
get_df_from_sql_file_with_params
read_query_from_file
fix_int_dtypes (used here in MySQLToS3Operator)
get_primary_keys (Snowflake/Redshift)
delete_all_objects_in_s3_bucket

And some other of the same style. What do you think?

Thanks

**Comment by potiuk** [comment, marked as answer]
What do you mean by "fit being in the code"? can you elaborate a bit please :)?

**Reply by JavierLopezT** [reply]
Yes, sorry. I mean opening a PR for including the functions in the Airflow code (either in providers or core) so that the functions are available to the people for use in their DAGs or python scripts. For example, if merged, you could import them as follow:

`from airflow.providers.amazon.aws.utils.s3 import copy_s3_firehose_to_another_bucket`

or 

`from airflow.utils.sql import get_df_from_sql_file_with_params`

Would that make sense?

**Reply by potiuk** [reply]
I think it is decision per-provider. Some providers already have some common utils or classes that can be reused. For example look here: https://github.com/apache/airflow/tree/main/airflow/providers/google/cloud/utils

We have some of those in Airflow as well:
https://github.com/apache/airflow/tree/main/airflow/utils

If you think you have a function to be good to add to providers, or core - feel free.
